scripts/gen_kobject_list.py

```diff
@@ -0,0 +1,443 @@
+
+        # Is this a kernel object, or a structure containing kernel objects?
+        if type_offset not in type_env:
+            continue
+
+        if "DW_AT_declaration" in die.attributes:
+            # FIXME: why does k_sys_work_q not resolve an address in the DWARF
+            # data??? Every single instance it finds is an extern definition
+            # but not the actual instance in system_work_q.c
+            # Is there something weird about how lib-y stuff is linked?
+            if name == "k_sys_work_q" and not work_q_found and name in syms:
+                addr = syms[name]
+                work_q_found = True
+            else:
+                continue
+        else:
+            if "DW_AT_location" not in die.attributes:
+                debug_die(die, "No location information for object '%s'; possibly stack allocated"
+                        % name)
+                continue
+
+            loc = die.attributes["DW_AT_location"]
+            if loc.form != "DW_FORM_exprloc":
+                debug_die(die, "kernel object '%s' unexpected location format" % name)
+                continue
+
+            opcode = loc.value[0]
+            if opcode != DW_OP_addr:
+
+                # Check if frame pointer offset DW_OP_fbreg
+                if opcode == DW_OP_fbreg:
+                    debug_die(die, "kernel object '%s' found on stack" % name)
+                else:
+                    debug_die(die, "kernel object '%s' unexpected exprloc opcode %s"
+                            % (name, hex(opcode)))
+                continue
+
+            addr = (loc.value[1] | (loc.value[2] << 8) | (loc.value[3] << 16) |
+                    (loc.value[4] << 24))
+
+        if addr < kram_start or addr >= kram_end:
+            if addr == 0:
+                # Never linked; gc-sections deleted it
+                continue
+
+            debug_die(die, "object '%s' found in invalid location %s" %
+                    (name, hex(addr)));
+            continue
+
+        type_obj = type_env[type_offset]
+        objs = type_obj.get_kobjects(addr)
+        all_objs.extend(objs)
+
+        debug("symbol '%s' at %s contains %d object(s)" % (name, hex(addr),
+            len(objs)))
+
+    debug("found %d kernel object instances total" % len(all_objs))
+    return all_objs
+
+
+header = """%compare-lengths
+%define lookup-function-name _k_object_lookup
+%language=ANSI-C
+%struct-type
+%{
+#include <kernel.h>
+#include <string.h>
+%}
+struct _k_object;
+%%
+"""
+
+
+# Different versions of gperf have different prototypes for the lookup function,
+# best to implement the wrapper here. The pointer value itself is turned into
+# a string, we told gperf to expect binary strings that are not NULL-terminated.
+footer = """%%
+struct _k_object *_k_object_find(void *obj)
+{
+    return _k_object_lookup((const char *)obj, sizeof(void *));
+}
+"""
+
+
+def write_gperf_table(fp, objs, static_begin, static_end):
```

Comment:
Played around with this, and it seems we can reduce the table even further if we do some processing on these pointers. The caveat is that the lookup will be slightly slower; but worth checking out, maybe later:

- Subtract the smaller pointer value from the list of all the pointers.  In the obj_validation test for qemu_x86, this made the largest value keep well below what a 16-bit integer can hold
- Look at all the values if they're aligned: if the least significant nibble for all the pointers is `0`, these values can be shifted-right by 2. In the obj_validation test for qemu_x86, all the values were well below what an 8-bit integer can hold

The _k_object structure has to be changed based on these findings, and _k_object_find() has to undo the compression, but this reduces the overhead without too much work